### PR TITLE
SDK-1356: version file usage and context

### PR DESCRIPTION
**Pull request by vol4onok**
## Issue ticket link
Ticket: https://spryker.atlassian.net/browse/SDK-1356.

## Checklist before requesting a review
- [ ] I have added ticket link to the PR description.
- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.

### CODE REVIEW

File: src/Infrastructure/EventListener/ContextOutputCommandListener.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+ */
+
+namespace SprykerSdk\Sdk\Infrastructure\EventListener;
+
+use SprykerSdk\Sdk\Core\Application\Dependency\ContextFactoryInterface;
+use SprykerSdk\SdkContracts\Entity\MessageInterface;
+use Symfony\Component\Console\Event\ConsoleTerminateEvent;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class ContextOutputCommandListener
```
Comment (sergeyspryker) [inline src/Infrastructure/EventListener/ContextOutputCommandListener.php:15]: In not sure that it should be there. I can see there are console specific things. May be move it into the presentation console layer?

File: src/Infrastructure/EventListener/ContextOutputCommandListener.php [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    /**
+     * @param \SprykerSdk\SdkContracts\Entity\MessageInterface $message
+     *
+     * @return string
+     */
+    protected function formatMessage(MessageInterface $message): string
+    {
+        $template = [
```
Comment (sergeyspryker) [inline src/Infrastructure/EventListener/ContextOutputCommandListener.php:63]: Move array with templates into the constant pls.

File: src/Infrastructure/Resources/config/services.yaml [hunk 3]
```diff
@@ -55,6 +55,16 @@ services:
                 event: console.error,
                 method: handle,
             }
+    command_context_output:
+        class: SprykerSdk\Sdk\Infrastructure\EventListener\ContextOutputCommandListener
+        arguments:
+            - "@context_factory"
+        tags:
+            - {
```
Comment (sergeyspryker) [inline src/Infrastructure/Resources/config/services.yaml:63]: Shouldn't use `{` `}` in yaml multiline syntax

File: src/Infrastructure/Service/LifecycleManager.php [hunk 4]
```diff
@@ -81,27 +89,25 @@ public function update(): void
     }
 
     /**
-     * @throws \SprykerSdk\Sdk\Infrastructure\Exception\SdkVersionNotFoundException
-     *
      * @return array<\SprykerSdk\SdkContracts\Entity\MessageInterface>
      */
     public function checkForUpdate(): array
     {
+        if ($this->environment !== 'prod') {
```
Comment (sergeyspryker) [inline src/Infrastructure/Service/LifecycleManager.php:96]: Why is version checking is prohibit for dev? Also we have `test` env for ci

File: src/Infrastructure/Service/LifecycleManager.php [hunk 5]
```diff
@@ -81,27 +89,25 @@ public function update(): void
...
+            return [new Message('Not possible to check updates for development version.')];
+        }
+
         $versionFilePath = $this->sdkDirectory . '/VERSION';
 
         if (!is_file($versionFilePath)) {
-            throw new SdkVersionNotFoundException(
-                sprintf('Could not find %s file, skip updatable check', $versionFilePath),
-            );
+            return [new Message('Could not find `VERSION` file, skip updatable check')];
```
Comment (sergeyspryker) [inline src/Infrastructure/Service/LifecycleManager.php:103]: Minor: move `VERSION` into the constant and use it in messages and in `$versionFilePath` var

File: src/Infrastructure/Service/LifecycleManager.php [hunk 6]
```diff
@@ -81,27 +89,25 @@ public function update(): void
...
         }
 
         $currentVersion = (string)file_get_contents($versionFilePath);
         $currentVersion = trim($currentVersion);
 
         if (!$currentVersion) {
-            throw new SdkVersionNotFoundException(
-                sprintf('Could not find version in the file "%s". File is empty.', $versionFilePath),
-            );
+            return [new Message('Could not find version in the file `VERSION`. File is empty.')];
```
Comment (sergeyspryker) [inline src/Infrastructure/Service/LifecycleManager.php:110]: Same

File: src/Infrastructure/EventListener/ContextOutputCommandListener.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+        if ($output->isVeryVerbose()) {
+            return MessageInterface::INFO;
+        }
+
+        if ($output->isDebug()) {
+            return MessageInterface::DEBUG;
+        }
+
+        return MessageInterface::ERROR;
```
Comment (sergeyspryker) [inline src/Infrastructure/EventListener/ContextOutputCommandListener.php:92]: Why is the normal verbosity is Error?

File: src/Infrastructure/EventListener/ContextOutputCommandListener.php [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    /**
+     * @param \Symfony\Component\Console\Output\OutputInterface $output
+     *
+     * @return int
+     */
+    protected function getVerbosity(OutputInterface $output): int
+    {
+        if ($output->isVerbose()) {
```
Comment (sergeyspryker) [inline src/Infrastructure/EventListener/ContextOutputCommandListener.php:80]: You can use mapping:
```
[
    OutputInterface::VERBOSITY_VERBOSE => MessageInterface::SUCCESS,
    .....
]
```

**Review by sergeyspryker — CHANGES_REQUESTED** [review]

File: bin/spryker-sdk.sh [hunk 9]
```diff
@@ -7,6 +7,11 @@ fi
 
 SDK_DIR="$(dirname $(dirname $0))"
 
+if [[ ! -f $SDK_DIR/VERSION ]]; then
+    echo "0.0.0" > $SDK_DIR/VERSION
+    exit 0
```
Comment (sergeyspryker) [inline bin/spryker-sdk.sh:12]: remove exit

**Review by sergeyspryker — CHANGES_REQUESTED** [review]

File: bin/spryker-sdk.sh (hunk 9 above)
Comment (sergeyspryker) [reply bin/spryker-sdk.sh:12]: git describe --abbrev=0 --tags > VERSION

File: src/Presentation/Console/Command/UpdateCommand.php [hunk 10]
```diff
@@ -52,35 +61,28 @@ public function __construct(
...
-        } catch (SdkVersionNotFoundException $exception) {
-            $output->writeln($exception->getMessage(), OutputInterface::VERBOSITY_VERBOSE);
-
-            return;
-        }
-
-        foreach ($messages as $message) {
-            $output->writeln($message->getMessage(), OutputInterface::VERBOSITY_VERBOSE);
+        foreach ($this->lifecycleManager->checkForUpdate() as $key => $message) {
+            $context->addMessage('check_updates_' . $key, $message);
```
Comment (Kyrylo-Meretskyi) [inline src/Presentation/Console/Command/UpdateCommand.php:85]: Move `check_updates_` to const

File: src/Infrastructure/EventListener/ContextOutputCommandListener.php (hunk 7 above)
Comment (vol4onok) [reply src/Infrastructure/EventListener/ContextOutputCommandListener.php:92]: I removed it. Checked only debug

File: src/Presentation/Console/Command/UpdateCommand.php (hunk 10 above)
Comment (vol4onok) [reply src/Presentation/Console/Command/UpdateCommand.php:85]: Don't need it. it's not const value.

File: bin/spryker-sdk.sh [hunk 11]
```diff
@@ -7,6 +7,10 @@ fi
 
 SDK_DIR="$(dirname $(dirname $0))"
 
+if [[ ! -f $SDK_DIR/VERSION ]]; then
+    git describe --abbrev=0 --tags > $SDK_DIR/VERSION
```
Comment (sergeyspryker) [inline bin/spryker-sdk.sh:11]: I guess it show version of target ptoject not sdk. Need to point sdk folder.

**Review by sergeyspryker — APPROVED** [review]

**Review by pavelmaksimov25 — APPROVED** [review]

File: src/Infrastructure/Service/LifecycleManager.php [hunk 12]
```diff
@@ -47,22 +48,30 @@ class LifecycleManager implements LifecycleManagerInterface
      */
     protected TaskYamlFileLoaderInterface $taskYamlFileLoader;
 
+    /**
+     * @var string
+     */
+    protected string $environment;
```
Comment (pavelmaksimov25) [inline src/Infrastructure/Service/LifecycleManager.php:54]: Not in use

File: src/Core/Application/Service/ContextFactory.php [hunk 13]
```diff
@@ -45,4 +45,12 @@ public function getContext(): ContextInterface
 
         return $this->context;
     }
+
+    /**
+     * @return bool
+     */
+    public function hasContext(): bool
```
Comment (pavelmaksimov25) [inline src/Core/Application/Service/ContextFactory.php:52]: Missed unit tests

**Review by pavelmaksimov25 — DISMISSED** [review]

**Review by pavelmaksimov25 — APPROVED** [review]

**Pull request merged**